scalafmt-dynamic/src/main/scala/org/scalafmt/dynamic/utils/ReentrantCache.scala

```diff
@@ -0,0 +1,61 @@
+package org.scalafmt.dynamic.utils
+
+import scala.annotation.tailrec
+import scala.concurrent.Await
+import scala.concurrent.Future
+import scala.concurrent.Promise
+import scala.concurrent.duration._
+import scala.util.Try
+
+class ReentrantCache[K, V] {
+  private[this] var cache: Map[K, Future[V]] = Map.empty
+
+  @tailrec
+  final def getOrAddToCache(key: K, shouldEvict: V => Boolean = _ => false)(get: () => V): V =
+    synchronized { // try to exit quickly from synchronized block
+      cache.get(key) match {
+        case Some(fut) => Right(fut)
+        case None =>
+          val p = Promise[V]()
+          cache += key -> p.future
+          Left(p)
+      }
+    } match {
+      case Right(fut) =>
+        /*fut.value match {
+          case Some(Success(value)) =>
+          //println(s"Value for $key already cached")
+          case Some(Success(value)) =>
+            //println(s"Previous calculation for $key failed")
+          case None =>
+            //println(s"Previous calculation for $key still ongoing")
+        }*/
+        val result = Await.result(fut, 1.minute) // get or wait for other thread to finish download
```

Comment:
Is this a 1 minute timeout limit to complete the download? If so, I think we need to increase the timeout since we can't expect everybody to have the same internet connection speed. I would go with a large value like 10 minutes, users see the download progress in the console output and can cancel the process with ctrl-c.